### PR TITLE
Password recovery via Steam

### LINKED ISSUE #393: Allow account recovery via Steam
Too many users

- forget their username,
- forget their password,
- forget their email and/or
- lose access to their email inbox

yet can still access their linked Steam account.
It would be nice if these people could 
- see their username and email of their FAF account via steam login
- reset their password via steam

### PR DESCRIPTION
**Pull request by Brutus5000**
fixes #393

### CODE REVIEW

File: src/main/java/com/faforever/api/user/SteamService.java [hunk 1]
```diff
@@ -66,4 +60,28 @@ boolean ownsForgedAlliance(String steamId) {
...
+    String recodedUri = builder.toUriString().replace("+", "%2B");
+    log.debug("Verification uri: {}", recodedUri);
+
+    // the Spring RestTemplate still struggles with the + character so we use the default Java http client
+    String result = HttpClient.newHttpClient()
+      .send(HttpRequest.newBuilder(new URI(recodedUri)).build(), BodyHandlers.ofString())
+      .body();
+
+    if (result == null || !result.contains("is_valid:true")) {
+      throw new IllegalArgumentException("Steam redirect could not be validated! Original response:\n" + result);
```
Comment (bukajsytlos) [inline src/main/java/com/faforever/api/user/SteamService.java:82]: Should this lead to internal server error?

File: src/main/java/com/faforever/api/user/UserService.java [hunk 2]
```diff
@@ -304,12 +319,66 @@ void performPasswordReset(String token, String newPassword) {
...
+    return userRepository.findOneBySteamId(steamId)
+      .map(steamUser ->
+        Pair.of(
+          steamUser,
+          fafTokenService.createToken(FafTokenType.PASSWORD_RESET,
+            Duration.ofSeconds(properties.getPasswordReset().getLinkExpirationSeconds()),
+            Map.of(UserService.KEY_USER_ID, String.valueOf(steamUser.getId())))
+        )
+      )
+      .map(steamUserAndTokenPair -> {
```
Comment (bukajsytlos) [inline src/main/java/com/faforever/api/user/UserService.java]: why not one .map?

File: src/main/java/com/faforever/api/user/UserService.java [hunk 3]
```diff
@@ -304,12 +319,66 @@ void performPasswordReset(String token, String newPassword) {
...
+          String token = steamUserAndTokenPair.getRight();
+
+          String callbackUrl = String.format(properties.getPasswordReset().getPasswordResetUrlFormat(), user.getLogin(), token);
+          return new CallbackResult(callbackUrl, List.of());
+        }
+      )
+      .orElseThrow(() -> ApiException.of(UNKNOWN_STEAM_ID, steamId));
+  }
+
+  CallbackResult performPasswordResetViaSteam(HttpServletRequest request, String token) {
```
Comment (bukajsytlos) [inline src/main/java/com/faforever/api/user/UserService.java]: I don't think this method is used now

File: src/main/java/com/faforever/api/user/UsersController.java [hunk 4]
```diff
@@ -51,86 +52,136 @@ public void register(HttpServletRequest request,
...
+                                           HttpServletResponse response) {
+    steamService.validateSteamRedirect(request);
+    String steamId = steamService.parseSteamIdFromLoginRedirect(request);
+    CallbackResult result = userService.requestPasswordResetViaSteam(steamId);
+    redirectCallbackResult(response, result);
+  }
+
+  @ApiOperation("Sets a new password for an account via Steam redirect.")
+  @GetMapping(path = "/performPasswordResetViaSteam", produces = APPLICATION_JSON_VALUE)
+  public void performPasswordResetViaSteam(HttpServletRequest request,
```
Comment (bukajsytlos) [inline src/main/java/com/faforever/api/user/UsersController.java]: also this is not used right? both password resets end in the same endpoint

File: src/test/java/com/faforever/api/user/UserServiceTest.java [hunk 5]
```diff
@@ -543,4 +556,50 @@ public void testResyncAccount() {
     verify(eventPublisher).publishEvent(any(UserUpdatedEvent.class));
   }
 
+  @Test
+  public void testPerformPasswordResetViaSteam(
```
Comment (bukajsytlos) [inline src/test/java/com/faforever/api/user/UserServiceTest.java]: also?

**Review by bukajsytlos — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/api/user/UserService.java (hunk 2 above)
Comment (Brutus5000) [reply src/main/java/com/faforever/api/user/UserService.java]: Good question. It made sense in earlier versions I guess ;) Will fix it

File: src/main/java/com/faforever/api/user/UserService.java (hunk 3 above)
Comment (Brutus5000) [reply src/main/java/com/faforever/api/user/UserService.java:354]: It's called from the UserController

Comment (Brutus5000) [reply src/main/java/com/faforever/api/user/UserService.java]: Yes, the new approach is much leaner.

File: src/main/java/com/faforever/api/user/UsersController.java (hunk 4 above)
Comment (Brutus5000) [reply src/main/java/com/faforever/api/user/UsersController.java]: Yes, the new approach is much leaner.

File: src/test/java/com/faforever/api/user/UserServiceTest.java (hunk 5 above)
Comment (Brutus5000) [reply src/test/java/com/faforever/api/user/UserServiceTest.java]: True

File: src/main/java/com/faforever/api/user/SteamService.java (hunk 1 above)
Comment (Brutus5000) [reply src/main/java/com/faforever/api/user/SteamService.java:82]: Probably not, but also nobody should ever encounter this error. For now I'll leave it as is.

**Pull request merged**